samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java

```diff
@@ -18,26 +18,323 @@
+      return wrapper;
+    }
+
+    TriggerHandlerImpl triggerHandler = new TriggerHandlerImpl();
+    TriggerImpl<M> triggerImpl = TriggerImpls.createTriggerImpl(trigger);
+    TriggerContextImpl triggerContext = new TriggerContextImpl(triggerKey);
+    wrapper = new TriggerImplWrapper(triggerKey, triggerImpl, triggerContext, triggerHandler);
+    triggers.put(triggerKey, wrapper);
+
+    return wrapper;
+  }
+
+  private void handleTrigger(TriggerKey<K> triggerKey, MessageCollector collector, TaskCoordinator coordinator) {
+    TriggerImplWrapper wrapper = triggers.get(triggerKey);
+
+    // Cancel all early triggers too when the default trigger fires.
+    if (triggerKey.getType() == TriggerType.DEFAULT) {
+      cancelTrigger(triggerKey);
+      cancelTrigger(new TriggerKey(TriggerType.EARLY, triggerKey.getKey()));
+    }
+
+    // Cancel non-repeating early triggers.
+    if (triggerKey.getType() == TriggerType.EARLY && !wrapper.isRepeating()) {
+      cancelTrigger(triggerKey);
+    }
+
+    WindowKey<K> windowKey = triggerKey.key;
+    WV wv = store.get(windowKey);
+
+    if (wv == null) {
+      return;
+    }
+
+    WindowPane<K, WV> paneOutput = new WindowPane<>(windowKey, wv, window.getAccumulationMode());
+    // Handle accumulation modes.
+    if (window.getAccumulationMode() == AccumulationMode.DISCARDING) {
+      store.put(windowKey, null);
+    }
+
+    if (paneOutput.getMessage() instanceof Collection) {
+      WV valCopy = (WV) new ArrayList<>((Collection<WV>) paneOutput.getMessage());
+      paneOutput = new WindowPane(windowKey, valCopy, window.getAccumulationMode());
+    }
+
+    WindowOperatorImpl.super.propagateResult((WM) paneOutput, collector, coordinator);
+  }
+
+  private void cancelTrigger(TriggerKey<K> triggerKey) {
+    TriggerImplWrapper wrapper = triggers.remove(triggerKey);
+    if (wrapper != null) {
+      wrapper.impl.cancel();
+    }
+  }
+
+  /**
+   * Implementation of the {@link TriggerContext} that allows {@link TriggerImpl}s to schedule and cancel callbacks.
+   */
+  public class TriggerContextImpl implements TriggerContext {
+    private final TriggerKey<K> triggerKey;
+
+    public TriggerContextImpl(TriggerKey<K> triggerKey) {
+      this.triggerKey = triggerKey;
+    }
+
+    @Override
+    public Cancellable scheduleCallback(Runnable runnable, long callbackTimeMs) {
+      TriggerTimerState timerState = new TriggerTimerState(triggerKey, runnable, callbackTimeMs);
+      pendingCallbacks.add(timerState);
+      return timerState;
+    }
+  }
+
+  private static class TriggerHandlerImpl implements TriggerImpl.TriggerCallbackHandler {
+
+    private boolean triggered;
+
+    @Override
+    public void onTrigger() {
+      triggered = true;
+    }
+
+    public boolean isTriggered() {
+      return triggered;
+    }
+
+    public void clearTrigger() {
+      triggered = false;
+    }
+  }
+
+  private class TriggerTimerState implements Comparable<TriggerTimerState>, Cancellable {
```

Comment:
Having a `Future` is not needed because we will not be using any other method in the `Future` API other than `cancel`. For instance, `get`s does not make much sense, as we would simply return `Void`.